apis/field_error.go

```diff
@@ -73,10 +73,84 @@ func (fe *FieldError) ViaField(prefix ...string) *FieldError {
 				strings.Join(append(prefix, oldPath), "."))
 		}
 	}
+
+	newPaths = flattenIndexes(newPaths)
+
 	fe.Paths = newPaths
 	return fe
 }
 
+// ViaIndex is used to attach an index to the next via field provided.
+// For example, if a type recursively validates a parameter that has a collection:
+//  for i, c := range spec.Collection {
+//    if err := doValidation(c); err != nil {
+//      return err.ViaIndex(i).ViaField("collection")
+//    }
+//  }
+func (fe *FieldError) ViaIndex(index ...int) *FieldError {
+	if fe == nil {
+		return nil
+	}
+
+	prefix := []string{}
+	for _, i := range index {
+		prefix = append(prefix, fmt.Sprintf("[%d]", i))
+	}
+
+	return fe.ViaField(prefix...)
+}
+
+// ViaKey is used to attach a key to the next via field provided.
+// For example, if a type recursively validates a parameter that has a collection:
+//  for k, v := range spec.Bag. {
+//    if err := doValidation(v); err != nil {
+//      return err.ViaKey(k).ViaField("bag")
+//    }
+//  }
+func (fe *FieldError) ViaKey(key ...string) *FieldError {
```

Comment:
Same as above, combine with `ViaField`?

Alternatively we could say `ViaFieldIndex(foo, bar)` as sugar for:
```
  return err.ViaIndex(bar).ViaField(foo)
```

Reply:
I intend to use this by using a helper function to validate a collection, so in my above example I have a simple case, in the real usage I have something like the following:
```
    if err := doCollectionValidation(spec.Collection); err != nil {
      return err.ViaField("collection")
    }
```

```
func doCollectionValidation(collection []Foos) {
   for i, c := range collection {
      if err := c.Validate(); err != nil {
        return err.ViaIndex(i)
      }
    }
}
```

`doCollectionValidation` is not aware of the field it is assigned to in the parent object, so making it a tuple does not help me in this way.